dali/operators/reader/loader/webdataset/tar_utils.cc

```diff
@@ -0,0 +1,207 @@
+  if (&other != this) {
+    stream_ = std::move(other.stream_);
+    std::tie(filename_, other.filename_) = std::forward_as_tuple(std::move(other.filename_), "");
+    std::tie(filesize_, other.filesize_) = std::forward_as_tuple(other.filesize_, 0);
+    std::tie(readoffset_, other.readoffset_) = std::forward_as_tuple(other.readoffset_, 0);
+    std::tie(archiveoffset_, other.archiveoffset_) = std::forward_as_tuple(other.archiveoffset_, 0);
+    std::tie(eof_, other.eof_) = std::forward_as_tuple(other.eof_, true);
+
+    if (instance_handle_ >= 0) {
+      Unregister(instance_handle_);
+    }
+    instance_handle_ = other.instance_handle_;
+  }
+  return *this;
+}
+
+uint64_t RoundToBlockSize(uint64_t count) {
+  return align_up(count, kBlockSize);
+}
+
+bool TarArchive::NextFile() {
+  if (eof_) {
+    return false;
+  }
+  Skip(RoundToBlockSize(filesize_) - readoffset_);
+  eof_ = ParseHeader();
+  return !eof_;
+}
+
+bool TarArchive::EndOfArchive() const {
+  return eof_;
+}
+
+std::string TarArchive::GetFileName() const {
+  return filename_;
+}
+
+uint64_t TarArchive::GetFileSize() const {
+  return filesize_;
+}
+
+std::shared_ptr<void> TarArchive::ReadFile() {
+  archiveoffset_ -= readoffset_;
+  int old_readoffset = readoffset_;
+  readoffset_ = 0;
+  stream_->Seek(archiveoffset_);
+  auto out = stream_->Get(filesize_);
+  if (out != nullptr) {
+    archiveoffset_ += filesize_;
+    readoffset_ = filesize_;
+  } else {
+    archiveoffset_ += old_readoffset;
+    readoffset_ = old_readoffset;
+  }
+  return out;
+}
+
+size_t TarArchive::Read(uint8_t* buffer, size_t count) {
+  if (eof_) {
+    return 0;
+  }
+  count = clamp(filesize_ - readoffset_, 0_u64, count);
+  uint64_t num_read_bytes = stream_->Read(buffer, count);
+  readoffset_ += num_read_bytes;
+  archiveoffset_ += num_read_bytes;
+  return num_read_bytes;
+}
+
+bool TarArchive::EndOfFile() const {
+  return readoffset_ >= filesize_;
+}
+
+inline void TarArchive::Skip(int64 count) {
+  stream_->Seek(archiveoffset_ += count);
+  readoffset_ += count;
+}
+
+int LibtarOpenTarArchive(const char*, int oflags, ...) {
+  va_list args;
+  va_start(args, oflags);
+  const int instance_handle_ = va_arg(args, int);
+  va_end(args);
+  return instance_handle_;
+}
+
+
+inline bool TarArchive::ParseHeader() {
+  TAR* handle;
+  tartype_t type = {LibtarOpenTarArchive, [](int) -> int { return 0; },
+                    [](int instance_handle_, void* buf, size_t count) -> ssize_t {
+                      const auto current_archive = instances[instance_handle_];
+                      const ssize_t num_read =
+                          current_archive->stream_->Read(reinterpret_cast<uint8_t*>(buf), count);
+                      current_archive->archiveoffset_ += num_read;
+                      return num_read;
+                    },
+                    [](int, const void*, size_t) -> ssize_t { return 0; }};
+  tar_open(&handle, "", &type, 0, instance_handle_, TAR_GNU);
```

Comment:
This is a bit wasteful - I think that this logic should be done once and here there should only be th_read.